lib/rust/parser/src/syntax/statement.rs

```diff
@@ -334,19 +341,164 @@ fn parse_statement<'s>(
-                content: apply_private_keywords(
-                    statement,
-                    items.drain(..),
-                    statement_context.visibility_context,
-                )
-                .map(StatementOrPrefix::Statement),
-            }
+        None => parse_expression_statement(
+            prefixes,
+            start,
+            item::Line { newline, items: mem::take(items) },
+            precedence,
+            statement_context.visibility_context,
+        )
+        .map_content(StatementOrPrefix::Statement),
+    }
+}
+
+fn take_trailing_newlines<'s>(
+    prefixes: &mut Vec<Line<'s, StatementPrefix<'s>>>,
+    trailing_newlines: &mut usize,
+    first_newline: &mut token::Newline<'s>,
+) -> Vec<token::Newline<'s>> {
+    let mut newlines = Vec::with_capacity(1 + *trailing_newlines);
+    newlines.extend((0..*trailing_newlines).map(|_| {
+        let Some(Line { newline, content: None }) = prefixes.pop() else { unreachable!() };
+        mem::replace(first_newline, newline)
+    }));
+    *trailing_newlines = 0;
+    newlines
+}
+
+fn take_doc_line<'s>(
+    prefixes: &mut Vec<Line<'s, StatementPrefix<'s>>>,
+    first_newline: &mut token::Newline<'s>,
+) -> Option<DocLine<'s>> {
+    let mut trailing_newlines = 0;
+    while let Some(prefix) = prefixes.get(prefixes.len().wrapping_sub(1) - trailing_newlines) {
+        let Some(content) = prefix.content.as_ref() else {
+            trailing_newlines += 1;
+            continue;
+        };
+        if let StatementPrefix::Documentation(_) = content {
+            let mut newlines =
+                take_trailing_newlines(prefixes, &mut trailing_newlines, first_newline);
+            let Some(Line { newline, content: Some(StatementPrefix::Documentation(docs)) }) =
+                prefixes.pop()
+            else {
+                unreachable!()
+            };
+            newlines.push(mem::replace(first_newline, newline));
+            newlines.reverse();
+            return Some(DocLine { docs, newlines });
+        } else {
+            break;
+        }
+    }
+    None
+}
+
+fn parse_expression_statement<'s>(
+    prefixes: &mut Vec<Line<'s, StatementPrefix<'s>>>,
+    start: usize,
+    mut line: item::Line<'s>,
+    precedence: &mut Precedence<'s>,
+    visibility_context: VisibilityContext,
+) -> Line<'s, Tree<'s>> {
+    let expression = precedence.resolve_offset(start, &mut line.items);
+    debug_assert!(line.items.len() <= start);
+    let expression = apply_private_keywords(expression, line.items.drain(..), visibility_context);
+    let mut first_newline = line.newline;
+    let expression =
+        expression.map(|expression| to_statement(prefixes, &mut first_newline, expression));
+    Line { newline: first_newline, content: expression }
+}
+
+fn to_statement<'s>(
+    prefixes: &mut Vec<Line<'s, StatementPrefix<'s>>>,
+    first_newline: &mut token::Newline<'s>,
+    expression_or_statement: Tree<'s>,
+) -> Tree<'s> {
```

Comment:
This `first_newline` is/becomes the first newline _after_ or _before_ the statement? It's quite hard (not impossible, but hard) to guess it from the code only.